### ISSUE TITLE
UpdateArray() function does not provide access to index of array's item

### ISSUE THREAD

**Issue by nihique** [MrWolfZ/ngrx-forms#46]
# UpdateArray() function does not provide access to index of array's item

Hi, I am working with form arrays and I have use case, that I want to sync certain property in array's item with its index (position in array). Problem is, that `updateArray(func)` func does not have index parameter, like I am used from JS `map((item, index) => ({...}))` function.

I have came up with workaround using `setValue()` for whole array, but I would prefer to add index as a third parameter to `updateArray()`. Is this possible?

```ts
const formReducerFn = (dateTimeSettings: DateTimeSettings) =>
    createFormGroupReducerWithUpdate<InvoiceFormValue>(
        {
            // this is how I would prefer to do it
            lineItems: updateArray<InvoiceLineItem>(lineItem => ({
                ...lineItem,
                sortOrder: 1, // should be 'index' instead of '1' - how to get it?
            })),
        },
        {
            // this is workaround I came up with
            lineItems: lineItems =>
                setValue(
                    lineItems.value.map((lineItem, index) => ({
                        ...lineItem,
                        sortOrder: index,
                    })),
                    lineItems,
                ),
        },
    );
```

**Comment by MrWolfZ** [comment]
This will be very easy to add. I'll try to work on it this week or at the latest on the weekend.

As a temporary workaround you can use the second parameter already provided for the update function (i.e. the parent array state itself) and then use `findIndex` on its `controls`, i.e. `sortOrder: arrayState.controls.findIndex(lineItem)`.

PS: I would recommend you use `setUserDefinedProperty` to store the sort order instead of storing it directly on the state since additional properties on the state will not be properly maintained by the library.

**Comment by nihique** [comment]
Thank you, that would be nice.

Btw, lineItem.sortOrder is part of entity, stored in our DB, so that is reason why it is there, otherwise I agree that it would be best to store it as you wrote...

**Comment by nihique** [comment]
I found, that the use case I wanted use does not work:

```ts
lineItems: updateArray<InvoiceLineItem>(lineItem => ({
                ...lineItem,
                sortOrder: lineItemsState.controls.findIndex(
                                x => x.id === lineItem.id,
                            ),
            })),
```
I rewrote it as this, but not sure if it is optimal way ;)
```ts
            lineItems: updateArray<InvoiceLineItem>(
                (lineItem, lineItemsState) =>
                    updateGroup<InvoiceLineItem>({
                        sortOrder: setValue(
                            lineItemsState.controls.findIndex(
                                x => x.id === lineItem.id,
                            ),
                        ),
                    })(lineItem),
            ),
```

**Comment by MrWolfZ** [comment]
Looks good to me.

**Comment by MrWolfZ** [comment]
I have just released version [3.0.0](https://github.com/MrWolfZ/ngrx-forms/releases/tag/v3.0.0) which contains a new update function `updateArrayWithFilter` that allows filtering on the index.
